### ISSUE TITLE
investigate db lag

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#125]
After implementing #85, there's a serious lag when chatting (and presumably when playing, too). It's disconcerting.

I already know that our JSONB encoding is expensive in terms of disk space, but I don't know enough about RDS to know why it's so slow. Tomorrow I'll need to hop on the Amazon console and try to figure out what's going on. I think we're running on the smallest instance, but I would have thought that for super trivial stuff that wouldn't matter. OTOH, when I added db access to logging in I noticed a lag and shrugged it off, so that suggests that the JSON encoding isn't the issue.

The lag doesn't appear on my development machine, but of course the DB I'm using is local (although I can configure my development machine to talk to our RDS instance).

**Comment by ctm** [comment]
I hopped on the RDS console and got into performance metrics. I've already disabled the errant code from #85, but once it's fixed tomorrow morning (and I've already fixed the panic), I can then figure out what's going on. It may be that having fsync turned on is what's doing it.

**Comment by ctm** [comment]
It may be that a t2.micro doesn't have enough memory to serve requests without paging. I've brought up a t3.small with Performance Monitoring enabled. I will do some tests locally and if that makes an improvement, I'll switch to it.

Since I'm spending a bit of time doing db work today anyway, I should also do #120, #133 and should also run my logging through a thread that operates in parallel with everything else so that there's no db-lag whatsoever for write-only (e.g. logging) operations (as opposed to table creation, where we can't proceed without the table-id).

**Comment by ctm** [comment]
Here are the things to do:

- [ ] test the t3.small (both interactive as well as some load testing)
- [ ] switch to the t3.small if it improves things
- [ ] store CborMinimal instead of JSONB #120 
- [ ] cards sent too often #123 
- [ ] don't have the db return rows when not needed
- [ ] have separate thread for write-only requests

Since I'm talking to potential investors, I not only need to eliminate the lag, but I should have some performance numbers.

**Comment by ctm** [comment]
FWIW, switching to t3.small didn't fix the lag, but I do have access to performance metrics and unless I'm misreading them, the stats don't show any lag at Amazon's side, which suggests that it's something I'm doing with Diesel that is causing the problems. I have some ideas.

It's annoying to have to track this down, however, this does suggest that we can go back to t2.micro once I figure out what's going on. OTOH, I haven't run yet today, so I should get out the door.

**Comment by ctm** [comment]
My guess is the issue is Nagle and the excess length of our packets (due to the JSON encoding) is what is biting us. If we use CborMinimal packets, Nagle might even help us if we have a thread for write-only requests, however my inclination is to simply disable Nagle completely, in part because my super vague recollection is that I had this same problem back in the multibot days and that was my solution. However, I just simply grabbed the socket and set the option myself. I don't yet see a trivial way to do that here, although it does look like I can get the raw socket file-descriptor  out of `pq-sys`.

I do see signs of nagle disabling in some of the sample rust net/web software I've downloaded, so I'll poke around in them a bit and see what they're calling to do the job. FWIW, I haven't noticed lag associated with my Websockets, so I'm assuming they're working fine, but I should look there too, since faster is better.

**Comment by ctm** [comment]
Turns out it was simply more round trips to the db server than I was expecting. The biggest surprise is that each time I ask for a connection from the pool it does what appears to be 2 round trips. However, there's still something going on which I don't understand that I'll further investigate after today's run. 

The upshot is that I can just not use the pool and that will get rid of a bunch of the lag, and once I fully understand what's going on I can first get the lag to be tolerable just by eliminating all the unnecessary round-trips, and then I can go up and do the various check-boxes above and that will completely eliminate the lag (due to using the thread to write the logs in parallel / after).

**Comment by ctm** [comment]
I did some more work on this yesterday, but the upshot is that I simply should stop most of this stuff and just make sure the db is co-located with the server even in this early stage of development. So, today I'll get back to playing with Convox and just bring up the minimal server and db instances needed to get some benchmarks.

**Comment by ctm** [comment]
I switched over to an EC2 instance and that will do for now. I didn't get any load testing done, but it's easy enough to spin up instances and deploy to them that I can do it sometime later on a separate ticket. As it is, this one is quite the embarrassment.
